modules/order/src/AvailabilityOrderProcessor.php

```diff
@@ -37,10 +38,23 @@ public function process(OrderInterface $order) {
     foreach ($order->getItems() as $order_item) {
       $purchased_entity = $order_item->getPurchasedEntity();
       if ($purchased_entity) {
-        $available = $this->availabilityManager->check($purchased_entity, $order_item->getQuantity(), $context);
-        if (!$available) {
-          $order->removeItem($order_item);
-          $order_item->delete();
+        /** @var \Drupal\commerce\AvailabilityResponseInterface $availability */
+        $availability = $this->availabilityManager->check($purchased_entity, $order_item->getQuantity(), $context);
+        if (!$availability->isNeutral()) {
+          if ($availability->getMax() == 0) {
+            $order->removeItem($order_item);
+            $order_item->delete();
+            drupal_set_message(t('The item %item is no longer available and has been removed from your cart.', [
+              '%item' => $order_item->getTitle(),
+            ]), 'warning');
+          }
+          elseif ($availability->getMax() < $order_item->getQuantity()) {
+            drupal_set_message(new PluralTranslatableMarkup(abs($availability->getMax() - $order_item->getQuantity()), '1 %item is currently unavailable. The quantity in your cart has been updated.', '@count %item are currently unavailable. The quantity in your cart has been updated.', [
+              '%item' => $order_item->getTitle(),
+            ]), 'warning');
+            $order_item->setQuantity($availability->getMax());
+            $order_item->save();
+          }
```

Comment:
With the use of the commerce_stock_backorderable module I posted in a sandbox, I seem to have a properly functioning backorder stock system while using this PR.  Seems to me the question should be submodule vs standard support in core stock, But maybe a review of my submodule would indicate if there are some inherent flaws to that approach as a general-purpose submodule vs a project-level custom module.